File: backend/src/main/java/com/kiwi/features/metrics/data/MetricsDataMapper.java
package com.kiwi.features.metrics.data;

import com.kiwi.common.types.PositiveOrZeroInteger;
import com.kiwi.features.users.data.UsersPersistence;

import java.time.LocalDate;

import static com.kiwi.common.utils.FormatUtils.formatDate;

public class MetricsDataMapper {

    public static MetricsDomain toDomain(MetricsDTO dto) {
        return (dto == null) ? null :
                new MetricsDomain(
                        LocalDate.parse(dto.getDate()),
                        new PositiveOrZeroInteger(dto.getMaxGoodTimeSeconds()),
                        new PositiveOrZeroInteger(dto.getCurrentGoodTimeSeconds()),
                        new PositiveOrZeroInteger(dto.getMaxBadTimeSeconds()),
                        new PositiveOrZeroInteger(dto.getCurrentBadTimeSeconds())
                );
    }

    public static MetricsDomain toDomain(MetricsPersistence persistence) {
        return (persistence == null) ? null :
                new MetricsDomain(
                        persistence.getDate(),
                        new PositiveOrZeroInteger(persistence.getMaxGoodTimeSeconds()),
                        new PositiveOrZeroInteger(persistence.getCurrentGoodTimeSeconds()),
                        new PositiveOrZeroInteger(persistence.getMaxBadTimeSeconds()),
                        new PositiveOrZeroInteger(persistence.getCurrentBadTimeSeconds())
                );
    }

    public static MetricsPersistence toPersistence(UsersPersistence usersPersistence, MetricsDomain domain) {
        return (domain == null) ? null :
                MetricsPersistence.builder()
                        .user(usersPersistence)
                        .date(domain.getDate())
                        .maxGoodTimeSeconds(domain.getMaxGoodTimeSeconds().value())
                        .currentGoodTimeSeconds(domain.getCurrentGoodTimeSeconds().value())
                        .maxBadTimeSeconds(domain.getMaxBadTimeSeconds().value())
                        .currentBadTimeSeconds(domain.getCurrentBadTimeSeconds().value())
                        .build();
    }

    public static MetricsPersistence toPersistence(UsersPersistence usersPersistence, MetricsDTO dto) {
        return toPersistence(usersPersistence, toDomain(dto));
    }

    public static MetricsDTO toDTO(MetricsDomain domain) {
        if (domain == null) return null;
        return new MetricsDTO(
                formatDate(domain.getDate()),
                domain.getMaxGoodTimeSeconds().value(),
                domain.getCurrentGoodTimeSeconds().value(),
                domain.getMaxBadTimeSeconds().value(),
                domain.getCurrentBadTimeSeconds().value()
        );
    }

    public static MetricsDTO toDTO(MetricsPersistence persistence) {
        return toDTO(toDomain(persistence));
    }
}

File: mobile/app/src/main/java/com/bellako/kiwi/features/metrics/data/MetricsDataMapper.kt
package com.bellako.kiwi.features.metrics.data

import android.os.Build
import androidx.annotation.RequiresApi
import com.bellako.kiwi.common.utils.DateUtils.dateToString
import com.bellako.kiwi.common.utils.DateUtils.stringToDate

object MetricsDataMapper {
    @RequiresApi(Build.VERSION_CODES.O)
    fun toDomain(dto: MetricsDTO): MetricsDomain =
        MetricsDomain(
            date = stringToDate(dto.date),
            maxGoodTimeSeconds = dto.maxGoodTimeSeconds,
            currentGoodTimeSeconds = dto.currentGoodTimeSeconds,
            maxBadTimeSeconds = dto.maxBadTimeSeconds,
            currentBadTimeSeconds = dto.currentBadTimeSeconds,
        )

    @RequiresApi(Build.VERSION_CODES.O)
    fun toDomain(state: MetricsState): MetricsDomain =
        MetricsDomain(
            date = stringToDate(state.date),
            maxGoodTimeSeconds = state.maxGoodTimeSeconds,
            currentGoodTimeSeconds = state.currentGoodTimeSeconds,
            maxBadTimeSeconds = state.maxBadTimeSeconds,
            currentBadTimeSeconds = state.currentBadTimeSeconds,
        )

    @RequiresApi(Build.VERSION_CODES.O)
    fun toState(domain: MetricsDomain): MetricsState =
        MetricsState(
            date = dateToString(domain.date),
            maxGoodTimeSeconds = domain.maxGoodTimeSeconds,
            currentGoodTimeSeconds = domain.currentGoodTimeSeconds,
            maxBadTimeSeconds = domain.maxBadTimeSeconds,
            currentBadTimeSeconds = domain.currentBadTimeSeconds,
        )

    @RequiresApi(Build.VERSION_CODES.O)
    fun toState(dto: MetricsDTO): MetricsState = toState(toDomain(dto))

    @RequiresApi(Build.VERSION_CODES.O)
    fun toDTO(domain: MetricsDomain): MetricsDTO =
        MetricsDTO(
            date = dateToString(domain.date),
            maxGoodTimeSeconds = domain.maxGoodTimeSeconds,
            currentGoodTimeSeconds = domain.currentGoodTimeSeconds,
            maxBadTimeSeconds = domain.maxBadTimeSeconds,
            currentBadTimeSeconds = domain.currentBadTimeSeconds,
        )

    @RequiresApi(Build.VERSION_CODES.O)
    fun toDTO(state: MetricsState): MetricsDTO = toDTO(toDomain(state))
}
